sdk/tools/whitelister/whitelister.cpp

```diff
@@ -0,0 +1,411 @@
+*/
+
+#include <cstdio>
+#include <cstdarg>
+#include <array>
+#include <fstream>
+#include <streambuf>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <vector>
+#include <direct.h>
+
+#define USE_WHITELIST
+#include "../include/reactos/wine/test.h"
+
+using namespace std;
+
+// Command to get whitelist entry:
+// reactos>git blame -L <Line>,<Line> --show-number --show-name <File Path>
+// e.g.:
+// E:\amd64\reactos>git blame -L 699,699 --show-number --show-name modules\rostests\apitests\advapi32\LockServiceDatabase.c
+// 8400fdc0786b rostests/apitests/advapi32/LockDatabase.c 701 (Thomas Faber 2012-06-29 11:48:35 +0000 699)     ok_err(ERROR_SUCCESS);
+//
+// To reverse it use
+// reactos>git blame -L <Old Line>,<Old Line>  -s --reverse --show-number --show-name <old hash> <old file>
+// e.g.
+// E:\amd64\reactos>git blame -L 701,701 --reverse 8400fdc0786b rostests\apitests\advapi32\LockDatabase.c
+// f2bc1f0e1118 modules/rostests/apitests/advapi32/LockServiceDatabase.c (Roman Masanin 2021-07-24 21:23:58 +0300 701)     ok_err(ERROR_SUCCESS);
+//
+
+#if 0
+#define Trace(...) fprintf(stderr, __VA_ARGS__)
+#else
+#define Trace(...)
+#endif
+
+struct WL_LINE_ENTRY
+{
+    int Line;
+    uint32_t Attributes;
+    string Reason;
+};
+
+struct WL_ENTRY
+{
+    string FullFileName;
+    string ShortFileName;
+    vector<WL_LINE_ENTRY> Lines;
+};
+
+string
+Execute(const char* cmd)
+{
+    array<char, 128> buffer;
+    string result;
+
+    Trace("Execute(\"%s\")\n", cmd);
+
+    std::unique_ptr<FILE, decltype(&_pclose)> pipe(_popen(cmd, "r"), _pclose);
+    if (pipe == nullptr)
+    {
+        throw std::runtime_error("popen() failed!");
+    }
+
+    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr)
+    {
+        result += buffer.data();
+    }
+
+    return result;
+}
+
+string
+ExecuteCmd(const char* Format, ...)
+{
+    array<char, _MAX_PATH> buffer;
+    va_list argptr;
+
+    va_start(argptr, Format);
+    _vsnprintf(buffer.data(), buffer.size(), Format, argptr);
+    va_end(argptr);
+
+    return Execute(buffer.data());
+}
+
+string
+git_blame(string &File, int Line)
+{
+    return ExecuteCmd("git blame -L %u,%u --show-number --show-name %s",
+                      Line, Line, File.c_str());
+}
+
+string
+git_rev_blame(string &OldHash, string &OldFile, int OldLine)
+{
+    return ExecuteCmd("git blame -L %u,%u -s --reverse --show-number --show-name %s %s",
+                      OldLine, OldLine, OldHash.c_str(), OldFile.c_str());
+}
+
+string
+git_rev_parse_head(void)
+{
+    string result = ExecuteCmd("git rev-parse @{0}");
+    return result.substr(0, 12);
+}
+
+bool
+git_diff_check(string &File)
+{
+    return ExecuteCmd("git diff --name-only %s", File.c_str()).length() != 0;
+}
+
+int
+AddWhitelistLine(const string &WhitelistFileName, const string &SourceFileName, int Line)
+{
+    return 0;
+}
+
+uint32_t
+GetWhiteListAttributesFromReason(string& Reason)
```

Comment:
how about comma-separated values ? `#WIN2003,#WIN8`

Reply:
Sure, you can do as you like, as long as the term is in there. "#WIN2003 and #WIN8" also works. Only "Works on #WIN7, but not on #WIN2003" wouldn't work as the text suggests.